### PR TITLE
Mechanisms and Components interface better with Mixtures

### PR DESCRIPTION
**Pull request by WilliamIX**
This PR is a draft, due to a couple of bugs that still need to be fixed. However, it is mostly done.

Bugs:
* @dr3y there is a bug in combinatorial promoter tests, in particular the last test which asks for a warning. I do not know where this comes from or why it doesnt work. 
* @mkratz there is a bug in your script for multi-tx functions (in examples \ specicialized component tutorials). Note that I have dramatically simplified your script and no longer have it redefine these Mechanisms because that is impossible to maintain.

Many changes regarding how Mechanisms and Components interface with Mixtures:

* Component.mechanisms is now a property with a setter. Component.add_mechanism(s) are the best functions to add new mechanisms to Components. Mechanisms are copied when added to a Component.

* Mixture.mechanisms is now a property with a setter. Mixture.add_mechanism(s) are the best functions to add new mechanisms to Mixtures. Mechanisms are copied when added to a Mixture.

*Mixture and Component mechanism dictionaries are no longer design to be accessed directly. Instead use Component.get_mechanism(mechanism_type). This allows for the removal of "default_mechanisms", versus "custom_mechanisms" versus "mechanisms" in both Component and Mixture. Now each only has Mechanisms. First the component mechanisms are searched, then the mixture mechanisms are searched.

* Mixture.components is now a property with a setter. Mixture.add_component(s) are the best functions for adding Components to a Mixture. Components are copied when added to a Mixture. Mixture.get_component will return a Component from inside the mixture using a couple of keywords.

* These new functions make subclassing mixtures and components easier. Now, in the construct of each, call super.__init__(...) then just add any mechanisms or components you want using add_mechanisms(...) or add_components(...), respectively. The example notebooks have been changed to reflect this.

* A number of bug were also fixed - including a bug in parameter defaulting!

### CODE REVIEW

**Comment by WilliamIX** [comment]
@zoltuz This PR is ready for review, and pending any changes merging. The coverage differences are due to deleting some large functions and replacing them with more compact methods, which resulting in a percentage drop in some areas even though overall I think most new features have come with thorough tests (and unit tests have been added for some old features as well).

File: Tests/test_mixtures_and_components_combinatorially.py [hunk 1]
```diff
@@ -108,8 +117,8 @@ def test_components_in_mixtures(self):
 					args["parameters"] = dict(self.parameters)
 					M = mixture(**args, components = [C])
 
-					if C not in M.components:
-						raise AttributeError(f"{C} not found in {M}.components")
+					if M.get_component(component = C) is None:
+						raise AttributeError(f"{C} not found in M.components={M.components}")
 
 					CRN = M.compile_crn()
```
Comment (ayush9pandey) [inline Tests/test_mixtures_and_components_combinatorially.py:123]: Once you merge #131, you'll get SBML validation here. Also, if there are other places that you are creating a CRN model it would be good if you could add that same SBML validation code.

Comment (WilliamIX) [reply Tests/test_mixtures_and_components_combinatorially.py:123]: I think this is the only place where CRNs are compiled - will make sure we get SBML validation.

**Comment by zoltuz** [comment]
@WilliamIX after merging PR #131 there are some conflicts to be resolved. Could you have a look at it?

**Comment by WilliamIX** [comment]
> @WilliamIX after merging PR #131 there are some conflicts to be resolved. Could you have a look at it?

Done.

File: biocrnpyler/component.py [hunk 2]
```diff
@@ -141,6 +121,13 @@ def add_attribute(self, attribute: str):
...
+        """Updates the ParameterDatabase inside a Component
+
+        Possible inputs:
+            parameter_file (string)
+            parameters (dict)
+            parameter_database (ParameterDatabase)        
+        """
 
         if parameter_file is not None:
             self.parameter_database.load_parameters_from_file.load_parameters_from_dictionary(parameter_file, overwrite_parameters = overwrite_parameters)
```
Comment (zoltuz) [inline biocrnpyler/component.py:133]: Not part of this PR, but this line looks fishy.

Comment (WilliamIX) [reply biocrnpyler/component.py:133]: good spot - fixed it

File: biocrnpyler/mechanisms_txtl.py [hunk 3]
```diff
@@ -368,19 +351,12 @@ class multi_tx(Mechanism):
...
             self.pol = pol
-
-        elif pol is None:
-            self.pol = Species("RNAP", material_type='protein')
         else:
-            raise ValueError("'pol' must be a string or Species")
-
+            raise ValueError("'pol' must be a Species")
 
         Mechanism.__init__(self, name=name, mechanism_type=mechanism_type, **keywords)
```
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/mechanisms_txtl.py:361]: fixed in both multi_tx and multi_tl.

File: biocrnpyler/mixture.py [hunk 4]
```diff
@@ -124,32 +103,209 @@ def set_species(self, species, material_type = None, attributes = None):
...
+            raise TypeError(f"mechanism keys must be strings. Recieved {mech_type}")
+
+        if mech_type in self._mechanisms and not overwrite:
+            raise ValueError(f"mech_type {mech_type} already in Mixture {self}. To overwrite, use keyword overwrite = True.")
+        else:
+            self._global_mechanisms[mech_type] = copy.deepcopy(mechanism)
+
     def update_parameters(self, parameter_file = None, parameters = None, overwrite_parameters = True):
         if parameter_file is not None:
             self.load_parameters_from_file.load_parameters_from_dictionary(parameter_file, overwrite_parameters = overwrite_parameters)
```
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/mixture.py:302]: fixed

File: biocrnpyler/mechanisms_txtl.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zoltuz) [reply biocrnpyler/mechanisms_txtl.py:361]: Thanks!

File: biocrnpyler/mixture.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/mixture.py:302]: confused that this isn't getting picked up by testing because I thought I had tests for loading functions...hmm

File: biocrnpyler/components_basic.py [hunk 5]
```diff
@@ -33,21 +33,13 @@ class DNA(Component):
...
-            mechanisms=None,  # custom mechanisms
-            parameters=None,  # customized parameters
-            attributes=None,
-            initial_conc=None,
-            parameter_warnings = True,
+            self, name: str, length=0, attributes = None,  # positional arguments
             **keywords
     ):
-        self.species = Species(name, material_type="dna",
+        self.species = self.set_species(name, material_type="dna",
```
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/components_basic.py:39]: self.set_species(...) is a general method used all the time. the "...)" was there, but on the line below. I have removed the enter, although likely the line will now be too long for pep8.

File: Tests/test_mixtures_and_components_combinatorially.py [hunk 6]
```diff
@@ -147,13 +167,13 @@ def test_dna_assemblies_with_combinations_of_mechanisms(self):
 				P = prom(**args)
 				for (rbs, args) in self.rbs_classes:
 					R = rbs(**args)
-					for mech_tx in self.transcription_mechs:
-						for mech_tl in self.translation_mechs:
-							Mtx = mech_tx()
-							Mtl = mech_tl()
+					for (mech_tx, args_tx) in self.transcription_mechs:
```
[the remark this answers is not in the document]
Comment (WilliamIX) [reply Tests/test_mixtures_and_components_combinatorially.py:170]: this code actually does both positional and keyword arguments grouped into one.

[the remark this answers is not in the document]
Comment (WilliamIX) [reply Tests/test_mixtures_and_components_combinatorially.py:170]: but if you really insist....

[the remark this answers is not in the document]
Comment (WilliamIX) [reply Tests/test_mixtures_and_components_combinatorially.py:170]: actually nevermind - the whole test uses args in like 20 places and its not worth changing.

File: biocrnpyler/mixture.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/mixture.py:302]: Oh right - this was one of the tests I couldn't get to work due to being unable to get pytest to deal with relative paths to load a parameter file.

[the remark this answers is not in the document]
Comment (zoltuz) [reply biocrnpyler/mixture.py:302]: you can use file mocking to create a fake file in memory and load it. There's an example like that in parameters.

File: Tests/test_mixtures_and_components_combinatorially.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zoltuz) [reply Tests/test_mixtures_and_components_combinatorially.py:170]: ok, we will take care of it in the next round.

File: biocrnpyler/components_basic.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zoltuz) [reply biocrnpyler/components_basic.py]: it's not about the line it's about a function is called set species and returns a value, why not setting up self.species inside the function?

File: biocrnpyler/mixture.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/mixture.py]: working on it.

File: biocrnpyler/components_basic.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/components_basic.py]: because that isn't what this function does. It is so Components can set any internal species using a string, component, or Species as an argument. It is used everywhere and has been for a long time.

[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/components_basic.py]: It does is not used for setting component.species.

[the remark this answers is not in the document]
Comment (zoltuz) [reply biocrnpyler/components_basic.py]: OK, I just try to find oddities during code review, it looked odd, but I see it now (sort of).

File: biocrnpyler/mixture.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zoltuz) [reply biocrnpyler/mixture.py]: Cool, keep me posted. After that everything is ready for merge.

[the remark this answers is not in the document]
Comment (WilliamIX) [reply biocrnpyler/mixture.py]: see slack
